backend/src/schema/resolvers/posts.js

```diff
@@ -211,10 +225,75 @@ export default {
       })
       return emoted
     },
+    pinPost: async (_parent, params, context, _resolveInfo) => {
+      let pinnedPostWithNestedAttributes
+      const { driver, user } = context
+      const session = driver.session()
+      const { id: userId } = user
+      let writeTxResultPromise = session.writeTransaction(async transaction => {
+        const deletePreviousRelationsResponse = await transaction.run(
+          `
+          MATCH (:User)-[previousRelations:PINNED]->(post:Post)
+          DELETE previousRelations
+          RETURN post
+        `,
+        )
+        return deletePreviousRelationsResponse.records.map(record => record.get('post').properties)
+      })
+      await writeTxResultPromise
+
+      writeTxResultPromise = session.writeTransaction(async transaction => {
+        const pinPostTransactionResponse = await transaction.run(
+          `
+            MATCH (user:User {id: $userId}) WHERE user.role = 'admin'
+            MATCH (post:Post {id: $params.id})
+            MERGE (user)-[pinned:PINNED {createdAt: toString(datetime())}]->(post)
+            RETURN post, pinned.createdAt as pinnedAt
+         `,
+          { userId, params },
+        )
+        return pinPostTransactionResponse.records.map(record => ({
+          pinnedPost: record.get('post').properties,
+          pinnedAt: record.get('pinnedAt'),
+        }))
+      })
+      try {
+        const [transactionResult] = await writeTxResultPromise
+        const { pinnedPost, pinnedAt } = transactionResult
+        pinnedPostWithNestedAttributes = {
+          ...pinnedPost,
+          pinnedAt,
+        }
+      } finally {
+        session.close()
+      }
+      return pinnedPostWithNestedAttributes
+    },
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-17T14:59:58Z" title="Thursday, October 17th 2019, 4:59:58 pm +02:00">Oct 17, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -131,15 +151,41 @@ export default {
         await session.run(cypherDeletePreviousRelations, { params })
 
         updatePostCypher += `
-          WITH post
           UNWIND $categoryIds AS categoryId
           MATCH (category:Category {id: categoryId})
           MERGE (post)-[:CATEGORIZED]->(category)
+          WITH post
         `
       }
 
+      if (unpinned) {
+        const cypherRemovePinnedStatus = `
+          MATCH ()-[previousRelations:PINNED]->(post:Post {id: $params.id})
+          DELETE previousRelations
+          RETURN post
+      `
+
+        await session.run(cypherRemovePinnedStatus, { params })
+      }
+
+      if (pinned) {
+        const cypherDeletePreviousRelations = `
+          MATCH ()-[previousRelations:PINNED]->(post:Post)
+          DELETE previousRelations
+          RETURN post
+        `
+
+        await session.run(cypherDeletePreviousRelations)
+
+        updatePostCypher += `
+          MATCH (user:User {id: $userId}) WHERE user.role = 'admin'
+          MERGE (user)-[:PINNED {createdAt: toString(datetime())}]->(post)
+          WITH post
+          `
+      }
```


This could be a little less complex, if we had different graphql mutations.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-17T22:46:30Z" title="Friday, October 18th 2019, 12:46:30 am +02:00">Oct 18, 2019</time>_

---


done